syntax = "proto3";

service raftImplemetation
{
    rpc RequestVote (VoteReq) returns (Vote);
    rpc AppendEntries (Empty) returns (Empty);
    rpc SendHeartBeat (Heartbeat) returns (AckHB);
    rpc ReplicateFile (ReplicateFileInfo) returns (Empty);
    rpc DeleteFile (ReplicateFileInfo) returns (Empty);
}

message ReplicateFileInfo
{
    string fileChunk = 1;
    string dcAddr = 2;
}

message Heartbeat
{
    string id = 1;
    int64 currentTerm = 2;
    Log log = 3;
}

message VoteReq
{
    string id = 1;
    int64 currentTerm = 2;
}


message Vote
{
    bool voted = 1;
    int64 currentTerm = 2;
}

message AckHB
{
    string ack = 1;
    repeated string dcAck = 2;
    map<string, int64> maxChunks = 3;
    int64 sizeAvail = 4;
}

message DcList
{
    repeated string dcs = 1;
}

message Log
{
    map<string, DcList> fileLog = 1;
    map<string, int64> maxChunks = 2;
    map<string, int64> dcSizes = 3;
}

message Empty
{

}

//python3 -m grpc.tools.protoc -I. --python_out=. --grpc_python_out=. raft.proto